In [ ]:
import pandas as pd

In [ ]:
import seaborn as sns

In [ ]:
corpus = pd.read_csv('/content/drive/MyDrive/BIA/Dataset/amazonreviews.tsv', sep='\t')

In [ ]:
corpus

label                                             review
0      pos  Stuning even for the non-gamer: This sound tra...
1      pos  The best soundtrack ever to anything.: I'm rea...
2      pos  Amazing!: This soundtrack is my favorite music...
3      pos  Excellent Soundtrack: I truly like this soundt...
4      pos  Remember, Pull Your Jaw Off The Floor After He...
...    ...                                                ...
9995   pos  A revelation of life in small town America in ...
9996   pos  Great biography of a very interesting journali...
9997   neg  Interesting Subject; Poor Presentation: You'd ...
9998   neg  Don't buy: The box looked used and it is obvio...
9999   pos  Beautiful Pen and Fast Delivery.: The pen was ...

[10000 rows x 2 columns]

In [ ]:
corpus.shape

(10000, 2)

In [ ]:
corpus.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   label   10000 non-null  object
 1   review  10000 non-null  object
dtypes: object(2)
memory usage: 156.4+ KB


In [ ]:
corpus['label'].value_counts()

label
neg    5097
pos    4903
Name: count, dtype: int64

In [ ]:
corpus.iloc[1]['review']

"The best soundtrack ever to anything.: I'm reading a lot of reviews saying that this is the best 'game soundtrack' and I figured that I'd write a review to disagree a bit. This in my opinino is Yasunori Mitsuda's ultimate masterpiece. The music is timeless and I'm been listening to it for years now and its beauty simply refuses to fade.The price tag on this is pretty staggering I must say, but if you are going to buy any cd for this much money, this is the only one that I feel would be worth every penny."

# Sentiment Analysis using VaderSentiment

In [ ]:
!pip install vaderSentiment # VADER (Valence Aware Dictionary and sEntiment Reasoner)

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 126.0/126.0 kB 8.8 MB/s eta 0:00:00


In [ ]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

In [ ]:
se = SentimentIntensityAnalyzer()

In [ ]:
se.polarity_scores(corpus.iloc[9997]['review'])

{'neg': 0.062, 'neu': 0.763, 'pos': 0.175, 'compound': 0.9467}

In [ ]:
s = 'The movie was worst movies i have ever seen'

In [ ]:
se.polarity_scores(s)

{'neg': 0.339, 'neu': 0.661, 'pos': 0.0, 'compound': -0.6249}

In [ ]:
corpus['Polarity Scores'] = corpus['review'].apply(lambda x: se.polarity_scores(x))

In [ ]:
corpus

label                                             review  \
0      pos  Stuning even for the non-gamer: This sound tra...   
1      pos  The best soundtrack ever to anything.: I'm rea...   
2      pos  Amazing!: This soundtrack is my favorite music...   
3      pos  Excellent Soundtrack: I truly like this soundt...   
4      pos  Remember, Pull Your Jaw Off The Floor After He...   
...    ...                                                ...   
9995   pos  A revelation of life in small town America in ...   
9996   pos  Great biography of a very interesting journali...   
9997   neg  Interesting Subject; Poor Presentation: You'd ...   
9998   neg  Don't buy: The box looked used and it is obvio...   
9999   pos  Beautiful Pen and Fast Delivery.: The pen was ...   

                                        Polarity Scores  
0     {'neg': 0.085, 'neu': 0.682, 'pos': 0.233, 'co...  
1     {'neg': 0.017, 'neu': 0.846, 'pos': 0.137, 'co...  
2     {'neg': 0.062, 'neu': 0.674, 'pos': 0.264, 'co...  
3     {'neg': 0.086, 'neu': 0.632, 'pos': 0.282, 'co...  
4     {'neg': 0.0, 'neu': 0.72, 'pos': 0.28, 'compou...  
...                                                 ...  
9995  {'neg': 0.017, 'neu': 0.852, 'pos': 0.132, 'co...  
9996  {'neg': 0.0, 'neu': 0.874, 'pos': 0.126, 'comp...  
9997  {'neg': 0.062, 'neu': 0.763, 'pos': 0.175, 'co...  
9998  {'neg': 0.087, 'neu': 0.913, 'pos': 0.0, 'comp...  
9999  {'neg': 0.027, 'neu': 0.821, 'pos': 0.152, 'co...  

[10000 rows x 3 columns]

In [ ]:
corpus['compound'] = corpus['Polarity Scores'].apply(lambda x: x['compound'])

In [ ]:
corpus

label                                             review  \
0      pos  Stuning even for the non-gamer: This sound tra...   
1      pos  The best soundtrack ever to anything.: I'm rea...   
2      pos  Amazing!: This soundtrack is my favorite music...   
3      pos  Excellent Soundtrack: I truly like this soundt...   
4      pos  Remember, Pull Your Jaw Off The Floor After He...   
...    ...                                                ...   
9995   pos  A revelation of life in small town America in ...   
9996   pos  Great biography of a very interesting journali...   
9997   neg  Interesting Subject; Poor Presentation: You'd ...   
9998   neg  Don't buy: The box looked used and it is obvio...   
9999   pos  Beautiful Pen and Fast Delivery.: The pen was ...   

                                        Polarity Scores  compound  
0     {'neg': 0.085, 'neu': 0.682, 'pos': 0.233, 'co...    0.9454  
1     {'neg': 0.017, 'neu': 0.846, 'pos': 0.137, 'co...    0.8957  
2     {'neg': 0.062, 'neu': 0.674, 'pos': 0.264, 'co...    0.9847  
3     {'neg': 0.086, 'neu': 0.632, 'pos': 0.282, 'co...    0.9814  
4     {'neg': 0.0, 'neu': 0.72, 'pos': 0.28, 'compou...    0.9840  
...                                                 ...       ...  
9995  {'neg': 0.017, 'neu': 0.852, 'pos': 0.132, 'co...    0.9610  
9996  {'neg': 0.0, 'neu': 0.874, 'pos': 0.126, 'comp...    0.9544  
9997  {'neg': 0.062, 'neu': 0.763, 'pos': 0.175, 'co...    0.9467  
9998  {'neg': 0.087, 'neu': 0.913, 'pos': 0.0, 'comp...   -0.3595  
9999  {'neg': 0.027, 'neu': 0.821, 'pos': 0.152, 'co...    0.9107  

[10000 rows x 4 columns]

In [ ]:
corpus['Sentiment']=corpus['compound'].apply(lambda x: 'pos' if x>=0 else 'neg')

In [ ]:
corpus

label                                             review  \
0      pos  Stuning even for the non-gamer: This sound tra...   
1      pos  The best soundtrack ever to anything.: I'm rea...   
2      pos  Amazing!: This soundtrack is my favorite music...   
3      pos  Excellent Soundtrack: I truly like this soundt...   
4      pos  Remember, Pull Your Jaw Off The Floor After He...   
...    ...                                                ...   
9995   pos  A revelation of life in small town America in ...   
9996   pos  Great biography of a very interesting journali...   
9997   neg  Interesting Subject; Poor Presentation: You'd ...   
9998   neg  Don't buy: The box looked used and it is obvio...   
9999   pos  Beautiful Pen and Fast Delivery.: The pen was ...   

                                        Polarity Scores  compound Sentiment  
0     {'neg': 0.085, 'neu': 0.682, 'pos': 0.233, 'co...    0.9454       pos  
1     {'neg': 0.017, 'neu': 0.846, 'pos': 0.137, 'co...    0.8957       pos  
2     {'neg': 0.062, 'neu': 0.674, 'pos': 0.264, 'co...    0.9847       pos  
3     {'neg': 0.086, 'neu': 0.632, 'pos': 0.282, 'co...    0.9814       pos  
4     {'neg': 0.0, 'neu': 0.72, 'pos': 0.28, 'compou...    0.9840       pos  
...                                                 ...       ...       ...  
9995  {'neg': 0.017, 'neu': 0.852, 'pos': 0.132, 'co...    0.9610       pos  
9996  {'neg': 0.0, 'neu': 0.874, 'pos': 0.126, 'comp...    0.9544       pos  
9997  {'neg': 0.062, 'neu': 0.763, 'pos': 0.175, 'co...    0.9467       pos  
9998  {'neg': 0.087, 'neu': 0.913, 'pos': 0.0, 'comp...   -0.3595       neg  
9999  {'neg': 0.027, 'neu': 0.821, 'pos': 0.152, 'co...    0.9107       pos  

[10000 rows x 5 columns]

In [ ]:
corpus.iloc[9997]

label                                                            neg
review             Interesting Subject; Poor Presentation: You'd ...
Polarity Scores    {'neg': 0.062, 'neu': 0.763, 'pos': 0.175, 'co...
compound                                                      0.9467
Sentiment                                                        pos
Name: 9997, dtype: object

In [ ]:
corpus.iloc[9997]['review']

"Interesting Subject; Poor Presentation: You'd be hard-pressed to tell a boring story about about a plucky country newspaper editor who had principles and stuck to 'em. But this author has done it. He says the book is an adaptation of his doctoral thesis. That fits. The book reads like a well-done, fully footnoted college paper, not like a dramatic novel. It strives to identify key historical facts at the expense of the reader's interest. With a modest amount of effort, it could have been reworked as a compelling autobiographical novel bolstered by its remarkable display of features of midwestern life during the first half of the 20th Century."

In [ ]:
corpus['label'].value_counts()

label
neg    5097
pos    4903
Name: count, dtype: int64

In [ ]:
corpus['Sentiment'].value_counts()

Sentiment
pos    6968
neg    3032
Name: count, dtype: int64

In [ ]:
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score

In [ ]:
y_true=corpus['label']
y_pred=corpus['Sentiment']

In [ ]:
confusion_matrix(y_true, y_pred)

array([[2629, 2468],
       [ 403, 4500]])

In [ ]:
print(classification_report(y_true, y_pred))

              precision    recall  f1-score   support

         neg       0.87      0.52      0.65      5097
         pos       0.65      0.92      0.76      4903

    accuracy                           0.71     10000
   macro avg       0.76      0.72      0.70     10000
weighted avg       0.76      0.71      0.70     10000



In [ ]:
print(accuracy_score(y_true, y_pred))

0.7129


Accuracy of Sentiment Analysis using vader sentiment is 71.29%

# Sentiment Analysis using Hugging face


In [ ]:
reviews = corpus['review'].tolist()

In [ ]:
reviews

['Stuning even for the non-gamer: This sound track was beautiful! It paints the senery in your mind so well I would recomend it even to people who hate vid. game music! I have played the game Chrono Cross but out of all of the games I have ever played it has the best music! It backs away from crude keyboarding and takes a fresher step with grate guitars and soulful orchestras. It would impress anyone who cares to listen! ^_^',
 "The best soundtrack ever to anything.: I'm reading a lot of reviews saying that this is the best 'game soundtrack' and I figured that I'd write a review to disagree a bit. This in my opinino is Yasunori Mitsuda's ultimate masterpiece. The music is timeless and I'm been listening to it for years now and its beauty simply refuses to fade.The price tag on this is pretty staggering I must say, but if you are going to buy any cd for this much money, this is the only one that I feel would be worth every penny.",
 'Amazing!: This soundtrack is my favorite music of all

In [ ]:
from transformers import pipeline

In [ ]:
# Load the sentiment analysis pipeline with a small efficient model
sentiment_pipeline = pipeline("sentiment-analysis", model="distilbert-base-uncased-finetuned-sst-2-english")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

Device set to use cpu


In [ ]:
# Perform sentiment analysis
results = sentiment_pipeline(reviews)

In [ ]:
results

[{'label': 'POSITIVE', 'score': 0.9993072748184204},
 {'label': 'POSITIVE', 'score': 0.9980586171150208},
 {'label': 'POSITIVE', 'score': 0.9998258948326111},
 {'label': 'POSITIVE', 'score': 0.9998522996902466},
 {'label': 'POSITIVE', 'score': 0.9997084736824036},
 {'label': 'POSITIVE', 'score': 0.9998525381088257},
 {'label': 'NEGATIVE', 'score': 0.999763548374176},
 {'label': 'POSITIVE', 'score': 0.9997997879981995},
 {'label': 'POSITIVE', 'score': 0.9986333250999451},
 {'label': 'POSITIVE', 'score': 0.9995325803756714},
 {'label': 'NEGATIVE', 'score': 0.9998093247413635},
 {'label': 'POSITIVE', 'score': 0.9990912675857544},
 {'label': 'POSITIVE', 'score': 0.999550998210907},
 {'label': 'NEGATIVE', 'score': 0.9923559427261353},
 {'label': 'NEGATIVE', 'score': 0.9997895359992981},
 {'label': 'NEGATIVE', 'score': 0.9995879530906677},
 {'label': 'POSITIVE', 'score': 0.9996565580368042},
 {'label': 'POSITIVE', 'score': 0.9820727109909058},
 {'label': 'NEGATIVE', 'score': 0.95782965421676

In [36]:
# Print results
for review, result in zip(reviews, results):
    print(f"Review: {review}\nSentiment: {result['label']}, (Confidence: {result['score']:.2f})\n")

Streaming output truncated to the last 5000 lines.
Sentiment: NEGATIVE, (Confidence: 1.00)

Review: Best Thing Ever Created!: My 3 year old is in LOVE with paint and this makes the whole idea of "Painting" more bearable for me. Instead of cleaning up paint messes on the carpet, on her and on the walls - I can sit her down on the sidewalk with this and she can paint away on everything (including the trees) to her heart's content. Wish they offered more colors, but being able to teach her that Blue+Yellow=Green is also a bonus.
Sentiment: POSITIVE, (Confidence: 0.99)

Review: doesn't wash off that easily: We bought this thinking our kids would love it. They did love painting. But the problem came when we tried to wash it off. We cleaned it about 30 minutes after the kids started and we had to get out the pressure washer. It took about an hour to clean up. It was by far the worst clean up. It does not come off that easily. the regular hose did nothing to get it off.
Sentiment: NEGATIVE, (

In [ ]:
corpus['SentimentHF'] = [result['label'] for result in results]

In [38]:
corpus.head()

label                                             review  \
0   pos  Stuning even for the non-gamer: This sound tra...   
1   pos  The best soundtrack ever to anything.: I'm rea...   
2   pos  Amazing!: This soundtrack is my favorite music...   
3   pos  Excellent Soundtrack: I truly like this soundt...   
4   pos  Remember, Pull Your Jaw Off The Floor After He...   

                                     Polarity Scores  compound Sentiment  \
0  {'neg': 0.085, 'neu': 0.682, 'pos': 0.233, 'co...    0.9454       pos   
1  {'neg': 0.017, 'neu': 0.846, 'pos': 0.137, 'co...    0.8957       pos   
2  {'neg': 0.062, 'neu': 0.674, 'pos': 0.264, 'co...    0.9847       pos   
3  {'neg': 0.086, 'neu': 0.632, 'pos': 0.282, 'co...    0.9814       pos   
4  {'neg': 0.0, 'neu': 0.72, 'pos': 0.28, 'compou...    0.9840       pos   

  SentimentHF  
0    POSITIVE  
1    POSITIVE  
2    POSITIVE  
3    POSITIVE  
4    POSITIVE

In [ ]:
corpus.iloc[:,-2:]

In [40]:
# Compare the two columns
corpus['Match'] = corpus['label'].map({'pos': 'POSITIVE', 'neg': 'NEGATIVE'}) == corpus['SentimentHF']

In [41]:
corpus['Match'].value_counts()

Match
True     9013
False     987
Name: count, dtype: int64

In [43]:
# Calculate the percentage of matches
match_percentage = corpus['Match'].mean() * 100
print(f"Percentage of matches: {match_percentage:.2f}%")

Percentage of matches: 90.13%


Accuracy of Sentiment Analysis using hugging face is 90.13%

# Conclusion

While both models demonstrate effectiveness, the accuracy comparison indicates that Hugging Face outperforms VADER in capturing nuanced sentiments, highlighting the importance of selecting appropriate tools for sentiment classification tasks. This analysis not only underscores the capabilities of advanced NLP techniques but also suggests avenues for further exploration in sentiment analysis methodologies.